dbos/system_database.go

```diff
@@ -2547,3 +2553,219 @@ func maskPassword(dbURL string) (string, error) {
+		}
+		return true
+	}
+
+	// PostgreSQL codes indicating connection/admin shutdown etc.
+	var pgerr *pgconn.PgError
+	if errors.As(err, &pgerr) {
+		switch pgerr.Code {
+		case pgerrcode.ConnectionException,
+			pgerrcode.ConnectionDoesNotExist,
+			pgerrcode.ConnectionFailure,
+			pgerrcode.SQLClientUnableToEstablishSQLConnection,
+			pgerrcode.SQLServerRejectedEstablishmentOfSQLConnection,
+			pgerrcode.AdminShutdown,
+			pgerrcode.CrashShutdown,
+			pgerrcode.CannotConnectNow:
+			return true
+		}
+	}
+
+	// pgx aggregate for connect attempts:
+	var cerr *pgconn.ConnectError
+	if errors.As(err, &cerr) {
+		return true
+	}
+
+	// Match most "connection closed" cases
+	if errors.Is(err, io.EOF) || strings.Contains(err.Error(), "conn closed") {
+		return true
+	}
+
+	// Net-level errors
+	var nerr net.Error
+	return errors.As(err, &nerr)
+}
+
+// retryConfig holds the configuration for a retry operation
+type retryConfig struct {
+	maxRetries     int // -1 for infinite retries
+	baseDelay      time.Duration
+	maxDelay       time.Duration
+	backoffFactor  float64
+	jitterMin      float64
+	jitterMax      float64
+	retryCondition func(error, *slog.Logger) bool
+	logger         *slog.Logger
+}
+
+// retryOption is a functional option for configuring retry behavior
+type retryOption func(*retryConfig)
+
+// withRetrierMaxRetries sets the maximum number of retry attempts (-1 for infinite)
+func withRetrierMaxRetries(maxRetries int) retryOption {
+	return func(c *retryConfig) {
+		c.maxRetries = maxRetries
+	}
+}
+
+// withRetrierBaseDelay sets the initial delay between retry attempts
+func withRetrierBaseDelay(delay time.Duration) retryOption {
+	return func(c *retryConfig) {
+		c.baseDelay = delay
+	}
+}
+
+// withRetrierMaxDelay sets the maximum delay between retry attempts
+func withRetrierMaxDelay(delay time.Duration) retryOption {
+	return func(c *retryConfig) {
+		c.maxDelay = delay
+	}
+}
+
+// withRetrierBackoffFactor sets the exponential backoff factor
+func withRetrierBackoffFactor(factor float64) retryOption {
+	return func(c *retryConfig) {
+		c.backoffFactor = factor
+	}
+}
+
+// withRetrierJitter sets the jitter range for retry delays (e.g., 0.95 to 1.05 for ±5% jitter)
+func withRetrierJitter(min, max float64) retryOption {
+	return func(c *retryConfig) {
+		c.jitterMin = min
+		c.jitterMax = max
+	}
+}
+
+// withRetrierCondition sets the function that determines if an error is retryable
+func withRetrierCondition(condition func(error, *slog.Logger) bool) retryOption {
+	return func(c *retryConfig) {
+		c.retryCondition = condition
+	}
+}
+
+// withRetrierLogger sets the logger for the retrier
+func withRetrierLogger(logger *slog.Logger) retryOption {
+	return func(c *retryConfig) {
+		c.logger = logger
+	}
+}
+
+// retry executes a function with retry logic using functional options
+func retry(ctx context.Context, fn func() error, options ...retryOption) error {
```

Comment:
It's not a huge deal, but this thing is massively overengineered, given that it's always called with the same parameters